src/server/logs/logs.go

```diff
@@ -103,12 +109,63 @@ func (ls LogService) GetLogs(ctx context.Context, request *logs.GetLogsRequest,
 						Log: &logs.LogMessage{
 							LogType: &logs.LogMessage_Verbatim{
 								Verbatim: &logs.VerbatimLogMessage{
-									Line: []byte(e.Line),
+									Line:      []byte(e.Line),
+									Timestamp: timestamppb.New(e.Timestamp),
+								},
+							},
+						},
+					},
+				}
+			case logs.LogFormat_LOG_FORMAT_PARSED_JSON:
+				resp = &logs.GetLogsResponse{
+					ResponseType: &logs.GetLogsResponse_Log{
+						Log: &logs.LogMessage{
+							LogType: &logs.LogMessage_Json{
+								Json: &logs.ParsedJSONLogMessage{
+									Verbatim: &logs.VerbatimLogMessage{
+										Line:      []byte(e.Line),
+										Timestamp: timestamppb.New(e.Timestamp),
+									},
+									NativeTimestamp: timestamppb.New(e.Timestamp),
 								},
 							},
 						},
 					},
 				}
+				jsonStruct := new(structpb.Struct)
+				if err := jsonStruct.UnmarshalJSON([]byte(e.Line)); err != nil {
```

Comment:
So, there needs to be some sophistication here to correct for broken Loki installs, which are common in the wild: https://github.com/pachyderm/pachyderm/blob/master/src/server/pps/server/api_server.go#L1780

Basically, we have to try and unmarshal and look for known fields.  If the fields are "log" and "timestamp", it needs to be unwrapped from docker's logging wrappper (unlikely as k8s doesn't use docker anymore).  If the line isn't JSON, then it's probably the CRI format, so that needs to be unwrapped and then reparsed.